### PR TITLE
Create experiment modularized FM SDK

### PR DESCRIPTION
**Pull request by zwu52**
FM Modularization. Part of go/firebase-next

Dependency updated: `app -> app-exp`, `app-type -> app-types-exp`, `messaging-types -> messaging-types-exp`
Dependency to-be-updates: firebase-installations (there is some issue with stubbing `getToken(installations)` call. Update dependency to firebase-installtions-exp in proceeding PR

### CODE REVIEW

**Review by Feiyang1 — COMMENTED** [review]
The API layer looks good, but we didn't really modularize the SDK as we still implements the methods on the controller. 

We need to move these methods to free floating functions in order to make them tree shakeable. I also think we should have a separate entry point for service worker, so we don't force people to include the service worker code in their client code.

**Comment by hsubox76** [comment]
FYI to avoid that changeset error you will want to add `@firebase/messaging-exp` to the `ignore` array in `.changeset/config.json`.

File: packages-exp/messaging-exp/package.json [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  },
+  "license": "Apache-2.0",
+  "peerDependencies": {
+    "@firebase/app-exp": "0.x",
+    "@firebase/app-types-exp": "0.x"
+  },
+  "dependencies": {
+    "@firebase/component": "0.1.19",
+    "@firebase/installations-exp": "0.x",
+    "@firebase/messaging-types-exp": "0.0.800",
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/package.json:36]:
```suggestion
    "@firebase/messaging-types-exp": "0.0.900",
```

File: packages-exp/messaging-exp/package.json [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    "type-check": "tsc --noEmit"
+  },
+  "license": "Apache-2.0",
+  "peerDependencies": {
+    "@firebase/app-exp": "0.x",
+    "@firebase/app-types-exp": "0.x"
+  },
+  "dependencies": {
+    "@firebase/component": "0.1.19",
+    "@firebase/installations-exp": "0.x",
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/package.json:35]:
```suggestion
    "@firebase/installations-exp": "0.0.900",
```

File: packages-exp/messaging-exp/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  interface FirebaseNamespace {
+    messaging: {
+      (app?: FirebaseApp): FirebaseMessaging;
+      isSupported(): boolean;
+    };
+  }
+  interface FirebaseApp {
+    messaging(): FirebaseMessaging;
+  }
+}
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/src/index.ts:49]: Can delete. We don't use namespace patching anymore.

File: packages-exp/messaging-exp/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+const messagingFactory: InstanceFactory<'messaging-exp'> = (
+  container: ComponentContainer
+) => {
+  if (!isSupported()) {
+    throw ERROR_FACTORY.create(ErrorCode.UNSUPPORTED_BROWSER);
+  }
+
+  return new MessagingService(
+    container.getProvider('app-exp').getImmediate(),
+    container.getProvider('installations-exp').getImmediate(),
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/src/index.ts:60]:
```suggestion
    container.getProvider('installations-exp-internal').getImmediate(),
```

File: packages-exp/messaging-exp/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    container.getProvider('analytics-internal')
+  );
+};
+
+_registerComponent(
+  new Component(
+    'messaging-exp',
+    messagingFactory,
+    ComponentType.PUBLIC
+  ).setServiceProps(NAMESPACE_EXPORTS)
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/src/index.ts:70]: Don't need `NAMESPACE_EXPORTS` anymore, since everything is exported directly.

File: packages-exp/messaging-exp/src/interfaces/internal-dependencies.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import { FirebaseApp } from '@firebase/app-types-exp';
+import { FirebaseInstallations } from '@firebase/installations-types-exp';
+import { Provider } from '@firebase/component';
+
+export interface FirebaseInternalDependencies {
+  app: FirebaseApp;
+  appConfig: AppConfig;
+  installations: FirebaseInstallations;
+  analyticsProvider: Provider<FirebaseAnalyticsInternalName>;
+}
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/src/interfaces/internal-dependencies.ts:29]:
```suggestion
import { _FirebaseInstallationsInternal } from '@firebase/installations-types-exp';
import { Provider } from '@firebase/component';

export interface FirebaseInternalDependencies {
  app: FirebaseApp;
  appConfig: AppConfig;
  installations: _FirebaseInstallationsInternal;
  analyticsProvider: Provider<FirebaseAnalyticsInternalName>;
}
```

File: packages-exp/messaging-types-exp/package.json [hunk 7]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "@firebase/messaging-types-exp",
+  "private": true,
+  "version": "0.0.800",
```
Comment (Feiyang1) [inline packages-exp/messaging-types-exp/package.json:4]:
```suggestion
  "version": "0.0.900",
```

File: packages-exp/messaging-exp/src/api/onBackgroundMessage.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+export function onBackgroundMessage(
+  messaging: MessagingService,
+  nextOrObserver: NextFn<MessagePayload> | Observer<MessagePayload>
+): Unsubscribe {
+  if (!!navigator) {
+    throw ERROR_FACTORY.create(ErrorCode.AVAILABLE_IN_SW);
+  }
+
+  // Initialize swController which resister listeners for onPush, onSubChange, onNotificationClick.
+  new SwController(messaging);
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/src/api/onBackgroundMessage.ts:34]: This feels like we can just use a function here since the newed object has no state and is never referenced.

File: packages-exp/messaging-exp/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  InstanceFactory
+} from '@firebase/component';
+import { ERROR_FACTORY, ErrorCode } from './util/errors';
+
+import { FirebaseMessaging } from '@firebase/messaging-types-exp';
+import { MessagingService } from './messaging-service';
+import { _FirebaseNamespace } from '@firebase/app-types/private';
+import { _registerComponent } from '@firebase/app-exp';
+import { isSupported } from './helpers/isSupported';
+
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/src/index.ts:31]:
```suggestion
export * from './api';
```

Comment (Feiyang1) [reply packages-exp/messaging-exp/src/index.ts:31]: Perhaps we should create 2 entry points, one for the client and one for the service worker. Then we only export the applicable APIs from them respectively.

**Review by Feiyang1 — COMMENTED** [review]
Overall it looks great! Left a few comments and I'd like to do a size analysis on the new APIs to see how treeshaking works once PR is merged.

**Comment by Feiyang1** [comment]
@zwu52 ping

File: packages-exp/messaging-exp/src/helpers/register.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  ComponentType,
+  InstanceFactory
+} from '@firebase/component';
+import { ERROR_FACTORY, ErrorCode } from '../util/errors';
+
+import { MessagingService } from '../messaging-service';
+import { _registerComponent } from '@firebase/app-exp';
+import { isSupported } from './isSupported';
+
+const messagingFactory: InstanceFactory<'messaging-exp'> = (
```
Comment (hsubox76) [inline packages-exp/messaging-exp/src/helpers/register.ts:30]: I see a CI build error:

```
[!] (plugin rpt2) Error: /home/runner/work/firebase-js-sdk/firebase-js-sdk/packages-exp/messaging-exp/src/helpers/register.ts(30,41): semantic error TS2344: Type '"messaging-exp"' does not satisfy the constraint 'never'.
src/helpers/register.ts
Error: /home/runner/work/firebase-js-sdk/firebase-js-sdk/packages-exp/messaging-exp/src/helpers/register.ts(30,41): semantic error TS2344: Type '"messaging-exp"' does not satisfy the constraint 'never'.
```

I'm not totally sure how it works, I see it is declared in `messaging-types-exp` but this file doesn't seem to know about it. Maybe it needs to import `@firebase/messaging-types-exp`? In storage-exp I declared it right inside index.ts:

https://github.com/firebase/firebase-js-sdk/blob/9fd3f5233077b45c5101789c427db51835484ce0/packages/storage/exp/index.ts#L105

File: common/api-review/messaging-exp.api.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+```ts
+
+import { FirebaseApp } from '@firebase/app-types-exp';
+import { FirebaseMessaging } from '@firebase/messaging-types-exp';
+import { MessagePayload } from '@firebase/messaging-types-exp';
+import { NextFn } from '@firebase/util';
+import { Observer } from '@firebase/util';
+import { Unsubscribe } from '@firebase/util';
+
+// @public (undocumented)
```
Comment (hsubox76) [inline common/api-review/messaging-exp.api.md:14]: Maybe you are planning to do this in another PR but to get documentation, the JSdoc comments need to be added in `src/api.ts`

**Review by hsubox76 — COMMENTED** [review]
I see a couple of unaddressed comments by Fei (or maybe I am looking at Github UI wrong). In addition to that I added one or two comments.

Also, you don't have to create a changeset for exp, but in order to stop the changeset tool from erroring when we run releases, you have to add the package to the ignore field in the changeset config: see comment https://github.com/firebase/firebase-js-sdk/pull/3822#issuecomment-729861348 and the error above it.

File: packages-exp/messaging-exp/src/messaging-service.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    this.firebaseDependencies = {
+      app,
+      appConfig,
+      installations,
+      analyticsProvider
+    };
+  }
+
+  _delete(): Promise<void> {
+    throw new Error('Method not implemented.');
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/src/messaging-service.ts:61]: Is it a TODO, or should it just be no-op?

**Review by Feiyang1 — COMMENTED** [review]
Looks great. I think we are almost ready. We need to create a release build for publishing to npm. You can take a look at https://github.com/firebase/firebase-js-sdk/blob/master/packages-exp/functions-exp/rollup.config.release.js. What it does is remove `-exp`s in the code, so the package is published as `@firebase/messaging` instead of `@firebase/messaging-exp` (it will be published under a special tag, so it won't interfere with the normal release).

We also need to add messaging to `firebase-exp`, so people can import from `firebase/messaging`.

In addition to that, we need to build 2 CDN scripts for messaging in `firebase-exp`, one for the client script, one for the service worker script. We can do it in a separate PR though. Let me know if you need help.

FYI, @hsubox76 - we need to make sure the service worker script is included in our distribution, otherwise people won't be able to use messaging.

**Comment by zwu52** [comment]
Thanks Fei & Christina for reviewing. 
I created a release rollup config.

in regards to "We also need to add messaging to firebase-exp, so people can import from firebase/messaging."  Looks like FM is already being imported in the constants.ts or is there another place to import?

**Comment by Feiyang1** [comment]
I meant that we need to make messaging-exp available in the umbrella package `firebase-exp` by adding a folder for messaging here - https://github.com/firebase/firebase-js-sdk/tree/master/packages-exp/firebase-exp

**Comment by zwu52** [comment]
Tried a bunch of fix for 
```
Error: /home/runner/work/firebase-js-sdk/firebase-js-sdk/packages-exp/messaging-exp/src/helpers/register.ts(38,27): semantic error TS2345: Argument of type '"app-exp"' is not assignable to parameter of type '"messaging-exp"'.
```

has this been seen previously?

**Comment by Feiyang1** [comment]
Make sure that the version of dependency `@firebase/component` is the same as `packages/component`. I got a different error when compiling, but it went away once I updated the version.

File: packages-exp/firebase-exp/package.json [hunk 13]
```diff
@@ -43,7 +43,9 @@
     "@firebase/functions-exp": "0.0.900",
     "@firebase/firestore": "2.0.5",
     "@firebase/performance-exp": "0.0.900",
-    "@firebase/remote-config-exp": "0.0.900"
+    "@firebase/remote-config-exp": "0.0.900",
+    "@firebase/messaging-exp": "0.0.900"
+
```
Comment (Feiyang1) [inline packages-exp/firebase-exp/package.json:48]: Can you please add `"messaging"` to the `components` array below?

File: packages-exp/messaging-exp/package.json [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    "@firebase/messaging-types-exp": "0.0.900",
+    "@firebase/util": "0.3.2",
+    "idb": "3.0.2",
+    "tslib": "^1.11.1"
+  },
+  "devDependencies": {
+    "rollup-plugin-json": "^4.0.0",
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/package.json:43]:
```suggestion
    "rollup-plugin-json": "4.0.0",
```

File: packages-exp/messaging-exp/package.json [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    "rollup-plugin-typescript2": "0.27.0",
+    "ts-essentials": "^6.0.7",
```
Comment (Feiyang1) [inline packages-exp/messaging-exp/package.json:45]:
```suggestion
    "ts-essentials": "7.0.1",
```

**Review by Feiyang1 — APPROVED** [review]
LGTM. One todo which we can do in a separate PR - we need to build the CDN script the service worker in `firebase-exp`, currently we only build a client CDN script for messaging.

**Comment by zwu52** [comment]
will do

**Comment by zwu52** [comment]
@hsubox76 @Feiyang1 
Trying to create the CDN build and need some input here. I think what I am targeting is to create a index.cdn.ts under pacakges-exp/firebase-exp/messaging/ and add a build rule in firebase-exp/rollup.config.js like that of app's (CMIIW). I am trying to have 
```
export { onBackgroudMessage } from '@firebase/messaging-exp';
```
for index.cdn.ts however `@firebase/messaging-exp` points to `messaging-exp/dist/index.d.ts`. What am I missing to have the file to point to `messaging-exp/dist/index.sw.d.ts`? (or am I working with the wrong materials).
